### PR TITLE
fix: 12861 trim unnecessary re-render in select widget in server side rendering

### PR DESCRIPTION
**Pull request by prsidhu**
## Description

Issue - onFilterChange is being triggered thrice on every character change.
Fix - 1. remove state variable to use filterText prop as is.
2. add debounce to onQueryChange method.
3. Blueprint is calling onQueryChange twice on every change. so added a check to compare filter values to triggered further functionality.

Fixes # (issue)

https://github.com/appsmithorg/appsmith/issues/12861

## Type of change

> Please delete options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: fix-12861 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 56.35 **(0.01)** | 37.75 **(0.01)** | 35.94 **(0)** | 56.57 **(0)**
 :green_circle: | app/client/src/selectors/commentsSelectors.ts | 85.25 **(1.64)** | 64.71 **(2.95)** | 73.33 **(0)** | 90.59 **(2.35)**
 :green_circle: | app/client/src/widgets/SelectWidget/component/index.tsx | 16.53 **(0.66)** | 0 **(0)** | 4.76 **(0.21)** | 17.27 **(0.75)**</details>

### CODE REVIEW

File: app/client/src/widgets/SelectWidget/component/index.tsx [hunk 1]
```diff
@@ -113,15 +110,10 @@ class SelectComponent extends React.Component<
     });
     return optionIndex === this.props.selectedIndex;
   };
-  onQueryChange = (filterValue: string) => {
-    this.setState({ query: filterValue });
+  onQueryChange = debounce((filterValue: string) => {
+    if (isEqual(filterValue, this.props.filterText)) return;
```
Comment (SatishGandham) [inline app/client/src/widgets/SelectWidget/component/index.tsx:114]: Is this the piece of code prevent the three calls?

Comment (prsidhu) [reply app/client/src/widgets/SelectWidget/component/index.tsx:114]: @SatishGandham  This prevents 1 call as onQueryChange method is being called twice by the blueprint component on each change.

File: app/client/src/widgets/SelectWidget/component/index.tsx [hunk 2]
```diff
@@ -113,15 +110,10 @@ class SelectComponent extends React.Component<
...
     this.props.onFilterChange(filterValue);
     this.listRef?.current?.scrollTo(0);
-    if (!this.props.serverSideFiltering) return;
-    return this.serverSideSearch(filterValue);
-  };
-  serverSideSearch = debounce((filterValue: string) => {
-    this.props.onFilterChange(filterValue);
   }, DEBOUNCE_TIMEOUT);
```
Comment (SatishGandham) [inline app/client/src/widgets/SelectWidget/component/index.tsx:117]: Why was this debounce not working earlier?

Comment (prsidhu) [reply app/client/src/widgets/SelectWidget/component/index.tsx:117]: @SatishGandham this debounce did work. The execute calls are triggered by onFilterChange which was already being called twice before this one.

**Review by SatishGandham — APPROVED** [review]

**Pull request merged**